### DISCUSSION TITLE
部署之后，在浏览器中寻址，将得到404

### DISCUSSION THREAD

**Discussion by xijaja** [solidjs/solid-router#208] [Q&A]
solid 和 solid-router 的版本是：

```
"@solidjs/router": "^0.5.1",
"solid-js": "^1.6.2"
```

我使用命令将其部署到了 deno

```
pnpm run build \
	&& cd dist \
	&& deployctl deploy --project=xxxxxx \
	--token=xxxxxxx \
	--prod https://deno.land/std@0.165.0/http/file_server.ts
```

然后我访问了项目的地址，在访问 “/” 根目录时，如果你没有登录，我将使用导航将访问者送到 “/login” 登录页：

```
const navigate = useNavigate();
navigate("/login");
```

此时访问的地址是：`https://domain/login`但我不能刷新，如果我点击浏览器的刷新按钮将得到 `Not Found`，
并且，我也不能输入其他访问地址比如 `https://domain/abcd` 即使我拥有这个路由，也会得到 `Not Found`，
同时，在控制台也将得到 `GET https://domain/login 404` 或 `GET https://domain/abcd 404` 的错误。

我不明白，是我的操作有问题吗？我应该修改哪些代码？

**Comment by xijaja** [comment]
我尝试使用了下面的依赖，在部署后，其结果是一致的：

```
"vite-plugin-solid": "^2.4.0",
"vite-plugin-windicss": "^1.8.8"
```

当访问了某个页面就不能刷新，一旦刷新页面就会得到 `Not Found`，这是否是一个 bug ？

**Comment by xijaja** [comment]
https://loud-goat-60.deno.dev

这个项目使用的是这个代码：https://github.com/solidjs/templates/tree/master/ts-router
你可以[点击进入](https://loud-goat-60.deno.dev) 然后点击顶部的 `about` 或 `error` ，然后刷新页面，你将立即得到 `Not Found`。

**Comment by Brendan-csel** [comment]
> When a page is visited, it cannot be refreshed, and once the page is refreshed, it will get it, is this a bug? Not Found

I don't have any experience with Deno sorry - but I don't think this is a `solid-router` issue.

It looks like Deno is just working as a plain web file server.  I imagine your `dist` folder only has one `.html` file in the root folder so that is the only page the Deno knows how to serve.  So Deno is returning 404 for all the other pages when you try to refresh them directly from the server.

My guess is you'd need to use a server-side framework that is compatible with Deno.  I'm not sure if that includes `solid-start` at the moment.

**Reply by xijaja** [reply]
那么应当如何部署呢？

尝试一下 netlify： https://delightful-gumption-e0f9cc.netlify.app

尝试一下vercel：https://solid-router-ts.vercel.app

他们均存在上述的问题，甚至我自己的服务器。

**Reply by Brendan-csel** [reply]
I think it depends on the host.

For example, a quick search about hosting SPAs on Netlify led me to [this article](https://www.netlify.com/blog/2020/04/07/creating-better-more-predictable-redirect-rules-for-spas/) which suggests adding `_redirects` file so all requests will get redirected (by Netlify) to your `index.html`.

**Reply by Brendan-csel** [reply]
Vercel seems to use `vercel.json` per [this article](https://medium.com/today-i-solved/deploy-spa-with-react-router-to-vercel-d10a6b2bfde8)

**Reply by xijaja** [reply]
OK ! 这确实是一个不错的办法！

我在 vercel 的部署中进行了尝试，现在 https://solid-router-ts.vercel.app/ 这个链接已经正常了！

但是，不论怎样，这都有点麻烦，并且让人意想不到。

**Reply by xijaja** [reply]
我喜欢 solidjs ，也许使用 solidstart 会更好，就不需要做些额外的工作了。

**Reply by xijaja** [reply]
像这个问题，Netlify 添加一个 `_redirects` 文件，Vercel 添加一个 `vercel.json` 文件，如果 deno 也添加一个什么文件，那么我使用自己的服务器部署，应该添加什么文件呢？文件中应该写点什么？

**Reply by Brendan-csel** [reply]
What are you using for a web server?  Express?

**Reply by xijaja** [reply]
Ubuntu 20.04 system, 
Caddy (similar to Nginx), 
golang fiber (http api)

**Reply by mosheduminer** [reply]
Since you are using caddy, you can set up redirects by adding this to your config (based on [this blog post](https://ma.ttias.be/set-up-custom-404-page-static-sites-caddy/)):
```
    handle_errors {
        @404 {
            expression {http.error.status_code} == 404
        }
        rewrite @404 /index.html
        file_server
    }
```
(I'm not actually familiar with caddy myself, I just found that blog post. The principle is to set the index.html as the 404 page).

**Reply by xijaja** [reply]
> Since you are using caddy, you can set up redirects by adding this to your config (based on [this blog post](https://ma.ttias.be/set-up-custom-404-page-static-sites-caddy/)):
> 
> ```
>     handle_errors {
>         @404 {
>             expression {http.error.status_code} == 404
>         }
>         rewrite @404 /index.html
>         file_server
>     }
> ```
> 
> (I'm not actually familiar with caddy myself, I just found that blog post. The principle is to set the index.html as the 404 page).

wow! That's awesome, it's effectively resolved.

**Comment by xijaja** [comment]
Seems like my problem has been solved, either on Vercel or my own server. But I still want to know, how should I do it in Deno Deploy. I tried Google but I didn't find an answer.

**Reply by mosheduminer** [reply]
I don't _think_ `std@0.165.0/http/file_server.ts` allows you to configure 404 behavior (based on reading [this page](https://deno.land/manual@v1.28.1/examples/file_server)). You could maybe open an issue with deno about that?

You could also create your own custom file server that serves up the index.html file instead of returning a 404 (you can probably just modify the official one).

**Reply by xijaja** [reply]
Okay, Thank u.

**Comment by xbmlz** [comment]
https://github.com/xbmlz/vitesse-solid，Opinionated Vite + SolidJS Starter Template
solid快速开发模板

**Reply by xijaja** [reply]
酷～

**Comment by xbmlz** [comment]
> solid 和 solid-router 的版本是：
> 
> ```
> "@solidjs/router": "^0.5.1",
> "solid-js": "^1.6.2"
> ```
> 
> 我使用命令将其部署到了 deno
> 
> ```
> pnpm run build \
> 	&& cd dist \
> 	&& deployctl deploy --project=xxxxxx \
> 	--token=xxxxxxx \
> 	--prod https://deno.land/std@0.165.0/http/file_server.ts
> ```
> 
> 然后我访问了项目的地址，在访问 “/” 根目录时，如果你没有登录，我将使用导航将访问者送到 “/login” 登录页：
> 
> ```
> const navigate = useNavigate();
> navigate("/login");
> ```
> 
> 此时访问的地址是：`https://domain/login`但我不能刷新，如果我点击浏览器的刷新按钮将得到 `Not Found`， 并且，我也不能输入其他访问地址比如 `https://domain/abcd` 即使我拥有这个路由，也会得到 `Not Found`， 同时，在控制台也将得到 `GET https://domain/login 404` 或 `GET https://domain/abcd 404` 的错误。
> 
> 我不明白，是我的操作有问题吗？我应该修改哪些代码？

[https://github.com/xbmlz/vitesse-solid，Opinionated](https://github.com/xbmlz/vitesse-solid%EF%BC%8COpinionated) Vite + SolidJS Starter Template
solid快速开发模板
